code/datums/chatmessage.dm

```diff
@@ -66,51 +59,93 @@
+		LAZYREMOVEASSOC(hearer.seen_messages, message_loc, src)
+
+		if(istype(hearers[hearer], /image))
+			hearer.images -= hearers[hearer]
+
+	if(!dropped_hash)
+		SSrunechat.messages_by_creation_string -= creation_parameters
+
 	message_loc = null
-	message = null
+
+	messages = null
+	hearers = null
+	approx_lines = null
+
 	return ..()
 
-/**
- * Calls qdel on the chatmessage when its parent is deleted, used to register qdel signal
- */
-/datum/chatmessage/proc/on_parent_qdel()
-	SIGNAL_HANDLER
+/datum/chatmessage/proc/end_of_life()
 	qdel(src)
 
+/datum/chatmessage/proc/on_hearer_qdel(datum/source)
+	SIGNAL_HANDLER
+	if(!istype(source, /client))
+		return
+
+	var/client/hearer = source
+	LAZYREMOVEASSOC(hearer.seen_messages, message_loc, src)
+	if(istype(hearers[hearer], /image))
+		var/image/seen_image = hearers[hearer]
+		messages -= seen_image
+
+	hearers -= hearer
+
 /**
- * Generates a chat message image representation
+ * generates the spanned text used for the final image and creates a callback in SSrunechat to call generate_image() next tick.
+ * This proc exists solely to handle everything in the image creation process before MeasureText() returns, as otherwise when the client
+ * returns the results of MeasureText() we are in the verb execution portion of the tick which means we're in danger of overtiming.
+ * delaying the final image processing to SSrunechat's next fire() fixes this.
  *
  * Arguments:
- * * text - The text content of the overlay
- * * target - The target atom to display the overlay at
- * * owner - The mob that owns this overlay, only this mob will be able to view it
- * * language - The language this message was spoken in
- * * extra_classes - Extra classes to apply to the span that holds the text
- * * lifespan - The lifespan of the message in deciseconds
+ * * text - the text used in the image, gets edited if it holds non allowed characters and/or the listener doesnt understand the speakers language
+ * * target - the atom creating the message being heard by others. the image we create will have its loc assigned to this atom
+ * * owner - the mob hearing the message from target, must have a client
+ * * lanugage - the language typepath this message is spoken in
+ * * extra_classes - the spans used for this message
  */
-/datum/chatmessage/proc/generate_image(text, atom/target, mob/owner, datum/language/language, list/extra_classes, lifespan)
+/datum/chatmessage/proc/prepare_text(text, atom/target, mob/owner, datum/language/language, list/extra_classes)
+	set waitfor = FALSE //because this waits on client input
 	/// Cached icons to show what language the user is speaking
 	var/static/list/language_icons
 
-	// Register client who owns this message
-	owned_by = owner.client
-	RegisterSignal(owned_by, COMSIG_PARENT_QDELETING, .proc/on_parent_qdel)
+	// Register client who hears this message
+	var/client/owned_by = owner.client
+	RegisterSignal(owned_by, COMSIG_PARENT_QDELETING, .proc/on_hearer_qdel)
```

Comment:
This runtimes on ghosts with the current TM

Reply:
God signals are depressing

Reply:
should be fixed from [`c92148e` (#65791)](https://github.com/tgstation/tgstation/pull/65791/commits/c92148ea793fd952ca0e403cde5a6663d8a207b3)